src_base/xeve_args.c

```diff
@@ -0,0 +1,2313 @@
+    }
+
+    if(opt->key != ARGS_NO_KEY)
+    {
+        snprintf(*help, ARGS_OPT_HELP_STRING_MAXLEN, "  -%c, --%s [%s]%s%s\n    : %s %s", opt->key, opt->key_long,
+                vtype, (optional) ? " (optional)" : "", (optional) ? default_value : "", opt->desc, allowed_values);
+    }
+    else
+    {
+        snprintf(*help, ARGS_OPT_HELP_STRING_MAXLEN, "  --%s [%s]%s%s\n    : %s %s", opt->key_long,
+                vtype, (optional) ? " (optional)" : "", (optional) ? default_value : "", opt->desc, allowed_values);
+    }
+
+
+
+    return XEVE_OK;
+}
+
+int xeve_args_num_options(ARGS_PARSER * args)
+{
+    if(!args) {
+        return 0;
+    }
+    return args->num_option;
+}
+
+/* Three-part iterator */
+char* xeve_args_first_key(ARGS_PARSER * args)             /* returns smallest element in n */
+{
+    ARGS_OPT *opt = NULL;
+
+    // do some checks
+    if( !args )
+        return NULL;
+
+    if(!args->num_option) return NULL;
+
+    opt = args->opts;
+    return opt->key_long;
+}
+
+// int nums_done(ARGS_PARSER * args, char* key);        /* returns 1 if key is past end */
+char* xeve_args_next_key(ARGS_PARSER * args, char* keyl)   /* returns next value after key */
+{
+    ARGS_OPT *opt = NULL;
+    int idx = 0;
+    int next_idx = 0;
+
+    // do some checks
+    if( !args)
+        return NULL;
+
+    if(keyl)
+    {
+        idx = args_search_long_key(args->opts, keyl);
+        if( idx  < 0 ) return NULL;
+
+        next_idx = idx + 1;
+
+        // if next idx is out of bound
+        if( next_idx > args->num_option-1) return NULL;
+
+
+    } else {
+        next_idx = 0;
+    }
+
+    opt = args->opts + next_idx;
+
+    return opt->key_long;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Internal function implementations
+///////////////////////////////////////////////////////////////////////////////
+
+static int args_search_long_key(ARGS_OPT * opts, const char * key)
+{
+    int oidx = 0;
+    ARGS_OPT * opt = NULL;
+
+    opt = opts;
+    while(opt->key != ARGS_END_KEY)
+    {
+        if(!strcmp(key, opt->key_long))
+        {
+            return oidx;
+        }
+        oidx++;
+        opt++;
+    }
+    return -1;
+}
+
+static int args_search_short_arg(ARGS_OPT * ops, const char key)
+{
+    int oidx = 0;
+    ARGS_OPT * o;
+
+    o = ops;
+
+    while(o->key != ARGS_END_KEY)
+    {
+        if(o->key != ARGS_NO_KEY && o->key == key)
+        {
+            return oidx;
+        }
+        oidx++;
+        o++;
+    }
+    return -1;
+}
+
+int args_read_value(ARGS_OPT * ops, const char * argv)
+{
+    if(argv == NULL || ops->opt_storage == NULL)
+    {
+        return -1;
+    }
+    if(argv[0] == '-' && (argv[1] < '0' || argv[1] > '9')) return -1;
+
+    switch(ops->value_type)
+    {
+        case VAL_TYPE_INT:
+            *((int*)ops->opt_storage) = atoi(argv);
+            break;
+
+        case VAL_TYPE_STRING:
+            strcpy((char*)ops->opt_storage, argv);
```

Comment:
This function is unsafe, use strncpy instead.

Reply:
> [<img alt="washish" width="40" height="40" src="https://avatars.githubusercontent.com/u/95278471?s=60&amp;v=4">](/washish)

x264.h
int x264_param_parse( x264_param_t *, const char *name, const char *value );